src/Nerdbank.Streams/MultiplexingStream.cs

```diff
@@ -1113,6 +1180,62 @@ private void OnChannelDisposed(Channel channel)
             }
         }
 
+        /// <summary>
+        /// Called when the local end was not able to completely write all the data to this channel due to an error,
+        /// leading to the transmission of a <see cref="ControlCode.ContentWritingError"/> frame being sent for this channel.
+        /// </summary>
+        /// <param name="channel">The channel whose writing was halted.</param>
+        /// <param name="exception">The exception that caused the writing to be haulted.</param>
+        private void OnChannelWritingError(Channel channel, Exception exception)
+        {
+            Requires.NotNull(channel, nameof(channel));
+
+            // Make sure that we are allowed to send error frames on this protocol version
+            if (this.protocolMajorVersion == 1)
+            {
+                if (this.TraceSource!.Switch.ShouldTrace(TraceEventType.Information))
+                {
+                    this.TraceSource.TraceEvent(
+                        TraceEventType.Information,
+                        (int)TraceEventId.WriteError,
+                        "Not informing remote side of write error on channel {0} since MultiplexingStream has protocol version of {1}",
+                        channel.QualifiedId,
+                        this.protocolMajorVersion);
+                }
+
+                return;
+            }
+
+            // Verify that we are able to communicate to the remote side on this channel
+            if (this.channelsPendingTermination.Contains(channel.QualifiedId) || !this.openChannels.ContainsKey(channel.QualifiedId))
+            {
+                if (this.TraceSource!.Switch.ShouldTrace(TraceEventType.Information))
+                {
+                    this.TraceSource.TraceEvent(
+                        TraceEventType.Information,
+                        (int)TraceEventId.WriteError,
+                        "Not informing remote side of write error on channel {0} as it is an improper state");
+                }
+
+                return;
+            }
+
+            // Create the payload to send to the remote side
+            WriteError error = new WriteError(exception.Message);
+            V2Formatter errorSerializationFormatter = (V2Formatter)this.formatter;
```

Comment:
Yeah I was not sure about this one because this relies on the fact that all future formatters must inherit from the V2Formatter for this code to work. By using the number, I imagined that by using the protocol numbers, we can prevent the need to have this requirement. I will currently change the code to have this check and we can revert the change back if necessary

Reply:
I am not sure about the other point that you make regarding doing the last part first. At least based on my understanding, we should always have error checking at the start of the method and I have generally thought to use the following structure: 
```
func() {
  if there is error{
           record the error
           return 
   }

   Actual function code
}
```

Reply:
> this relies on the fact that all future formatters must inherit from the V2Formatter for this code to work

Oh, that's a good point. With your model, we'd get a InvalidCastException if the assumption doesn't hold, whereas with my suggestion it would quietly take the non-supporting branch. I actually like your idea better now.

As far as error checking before actual function code, that's good for argument validation. But when it comes to taking two paths based on _state_ (fields), it's not nearly as set in stone. We often use pattern like this:

```
if state1:
   ....
else if state 2:
   .....
else:
   throw badstate; // or return null;
```

To write this with `throw badstate` on top would require redundant condition checks, and the compiler would no longer be able to help us enforce that the if chain or switch statement later on was 'complete' and guaranteed to provide a branch for every possible case.

Reply:
I see - I think that certainly makes sense. The one thing to note that we have two different cases where we don't send the error to the remote side and they in theory should be dealt with differently. The first case is that the user has called this method when it is using a non-supported protocol version, such as protocol version 1, in which case we log that this is the case and return without throwing the error. On the other hand, if the channel is in a state where we can't send this message then we throw an error because there is a larger issue with the system